server/src-lib/Hasura/Server/Auth/JWT.hs

```diff
@@ -80,36 +80,66 @@ defaultRoleClaim = "x-hasura-default-role"
 defaultClaimsNamespace :: Text
 defaultClaimsNamespace = "https://hasura.io/jwt/claims"
 
-type CustomClaimsMap = Map.HashMap Text J.JSONPath
+data JWTClaimsMapField = JWTClaimsMapJSONPath !J.JSONPath
+                       | JWTClaimsMapString !Text
+                       | JWTClaimsMapAllowedRoles ![Text]
+                       deriving (Eq,Show)
+
+instance J.ToJSON JWTClaimsMapField where
+  toJSON (JWTClaimsMapJSONPath jsonPath) =
+    J.Object $ Map.fromList $ [("path", J.String $ T.pack . encodeJSONPath $ jsonPath)]
+  toJSON (JWTClaimsMapString s) = J.String s
+  toJSON (JWTClaimsMapAllowedRoles allowedRoles) = J.toJSON allowedRoles
+
+type CustomClaimsMap = Map.HashMap Text JWTClaimsMapField
 
 data JWTClaimsMap
   = JWTClaimsMap
-  { jcmDefaultRole  :: !J.JSONPath
-  , jcmAllowedRoles :: !J.JSONPath
+  { jcmDefaultRole  :: !JWTClaimsMapField
+  , jcmAllowedRoles :: !JWTClaimsMapField
   , jcmCustomClaims :: !CustomClaimsMap
   } deriving (Show, Eq)
```

Comment:
Assuming `JWTClaimsMapField` -> `JWTClaimsMapValue`

We should avoid `throw500` at https://github.com/rakeshkky/graphql-engine/pull/4/files#diff-ea85db56216e10e4ed52755329ec3850R403. The `jcmDefaultRole` and values in `CustomClaimsMap` are either `JSONPath with maybe default` or `Static string`. Only `jcmAllowedRole` is either `JSONPath with maybe default` or `Static string array`.  We need to parameterize `JWTClaimsMapValue`  

```haskell
data JWTClaimsMapValueG v
  = JWTClaimsJSONPath !J.JSONPath !(Maybe v)
  | JWTClaimsStatic !v
  deriving (Show, Eq, Foldable)

instance (J.FromJSON v) => J.FromJSON (JWTClaimsMapValueG v) where
  parseJSON (J.Object obj) = undefined -- parse as JWTClaimsJSONPath {"path": "<jsonpath>", "default": "v"}
  parseJSON v              = JWTClaimsStatic <$> J.parseJSON v

type JWTClaimsMapValue = JWTClaimsMapValueG Text
type JWTClaimsAllowedRoles = JWTClaimsMapValueG [Text]

type CustomClaimsMap = Map.HashMap SessionVariable JWTClaimsMapValue

data JWTClaimsMap
  = JWTClaimsMap
  { jcmDefaultRole  :: !JWTClaimsMapValue
  , jcmAllowedRoles :: !JWTClaimsAllowedRoles
  , jcmCustomClaims :: !CustomClaimsMap
  } deriving (Show, Eq)

instance J.FromJSON JWTClaimsMap where
  parseJSON _ = do
    -- 1. Parse as object like "HashMap SessionVariable (JWTClaimsMapValueG J.Value)"
    -- 2. For x-hasura-allowed-roles key, case match parameterised J.Value with J.Array [J.String t] using fmap
    -- 3. For x-hasura-default-role key and other keys, case match parameterised J.Value with J.String using fmap
    undefined
```